File: python/394.py
class Solution:
    def decodeString(self, s: str) -> str:
        def helper(s):
            j = 0
            while(j < len(s) and s[j].isalpha()):
                j += 1
            return s[:j],int(s[j:])
            
        cur = ''
        stack = []
        for i in s:
            if i == '[':
                stack.append(cur)
                cur = ''
            elif i == ']':
                last = stack.pop()
                lastChar,lastNum = helper(last)
                # print(lastChar,lastNum)
                cur = lastChar + lastNum*cur
                # print(cur)
            else:
                cur += i
        return cur
class Solution:
    def decodeString(self, s: str) -> str:
        stack = []
        num = []
        j = 0
        cur = ''
        for ch in s:
            if ch == '[':
                num.append(j)
                j = 0
                stack.append(cur)
                cur = ''
            elif ch == ']':
                S = stack.pop()
                q = num.pop()
                cur = S + cur * q
            elif ch.isnumeric():
                j = int(ch) + 10 * j
            else:
                cur += ch
        return cur

File: java/394.java
class Solution {
    public String decodeString(String s) {
        Stack<String>stack = new Stack();
        Stack<Integer> num = new Stack();
        StringBuilder cur = new StringBuilder();
        int j = 0;
        for(char k:s.toCharArray()){
            if(k == '['){
                num.push(j);
                j = 0;
                stack.push(cur.toString());
                cur = new StringBuilder();
            }else if(k == ']'){
                String S = stack.pop();
                // System.out.println(S);
                int q = num.pop();
                // System.out.println(q);
                String temp = cur.toString();
                cur = new StringBuilder();
                cur.append(S);
                while(q > 0){
                    cur.append(temp);
                    q --;
                }
                // System.out.println("cur = "+cur);
            }else if(k >= '0' && k <= '9'){
                j = j * 10 + k - '0';
                // num.push(j);
            }else{
                cur.append(k);
            }
        }
        return cur.toString();
    }
    
    
}